Playwright test workflow migration from Blocks repo

lanej0 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Migrates the Playwright-based e2e test workflow from the Blocks repository and sets to run on Blocks PRs.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1.
2.
3.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Commit:
First pass at Playwright test migration

Commit:
Set working directory

Commit:
Remove exec

Commit:
Configure the environment before testing

Commit:
Remove filter

Commit:
Update paths for tests

Commit:
Try explicitly stating working-directory

Commit:
Change tests directory

Commit:
Try running all tests

Commit:
Uncommented project

Commit:
Update commands to run tests

Commit:
Comment out env config

Commit:
Uncomment

lanej0 commented:
@gigitux I think this is ready to go.  I've got some tests failing though -- so I'm not sure if I've missed something in terms of setup/configuration, or if there's just some failures.  Let me know if I've missed something.

Commit:
Changelog

tarhi-saad commented on .github/workflows/blocks-playwright.yml:
```diff
@@ -0,0 +1,66 @@
+              run: pnpm --filter='@woocommerce/block-library' wp-env start
+
+            - name: Configure the environment
+              run: pnpm wp-env:config
```
@lanej0, we use different configs for Playwright. Playwright configs live in `./tests/e2e/bin` (in Woo Blocks). First we should update the mapping in `plugins/woocommerce-blocks/.wp-env.json`:

```diff
"env": {
	"tests": {
		"mappings": {
			"wp-content/mu-plugins": "./node_modules/@wordpress/e2e-tests/mu-plugins",
			"wp-content/plugins/gutenberg-test-plugins": "./node_modules/@wordpress/e2e-tests/plugins",
			"wp-cli.yml": "./wp-cli.yml",
			"custom-plugins": "./tests/e2e/mocks/custom-plugins",
-			"wp-content/plugins/woocommerce/blocks-bin": "./bin"
+			"wp-content/plugins/woocommerce/blocks-bin": "./bin",
+			"wp-content/plugins/woocommerce/blocks-bin/playwright": "./tests/e2e/bin"
		}
	}
},
```

Then we should fix the path in `plugins/woocommerce-blocks/tests/e2e/bin/test-env-setup.sh`

```diff
#!/usr/bin/env bash

# Extract the relative path from the plugin root to this script directory.
# By doing so, we can run this script from anywhere.
script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
head_dir=$(cd "$(dirname "$script_dir")" && cd ../../.. && pwd)
relative_path=${script_dir#$head_dir/}

# Run the main script in the container for better performance.
-wp-env run tests-cli -- bash wp-content/plugins/$relative_path/scripts/index.sh
+wp-env run tests-cli -- bash wp-content/plugins/woocommerce/blocks-bin/playwright/scripts/index.sh
```

Finally, let's update the `env:start` script in `plugins/woocommerce-blocks/package.json`

```diff
-"env:start": "pnpm run wp-env start && ./tests/e2e/bin/test-env-setup.sh",
+"env:start": "pnpm --filter='@woocommerce/block-library' wp-env start && ./tests/e2e/bin/test-env-setup.sh",
```
`.github/workflows/blocks-playwright.yml` can be updated as such:

```diff
steps:
            - uses: actions/checkout@v3

            - name: Setup WooCommerce Monorepo
              uses: ./.github/actions/setup-woocommerce-monorepo
              with:
                  install: '@woocommerce/plugin-woocommerce...'
                  build: '@woocommerce/plugin-woocommerce'

            - name: Install Playwright
              run: pnpm --filter='@woocommerce/block-library' exec playwright install --with-deps

-          - name: Start wp-env
+          - name: Start wp-env and configure the environment
+            run: pnpm --filter='@woocommerce/block-library' env:start
-            run: pnpm --filter='@woocommerce/block-library' wp-env start

-            - name: Configure the environment
-              run: pnpm wp-env:config
```

tarhi-saad reviewed (changes requested):
Thank you, @lanej0, for working on this PR! 🙌 

> I've got some tests failing though -- so I'm not sure if I've missed something in terms of setup/configuration, or if there's just some failures. Let me know if I've missed something.

I wrote a comment below regarding the setup. But we also have some failing tests:

The `Add to cart` label was changed recently, so the function `addToCart()` fails in our tests. To fix it you can update the file `plugins/woocommerce-blocks/tests/e2e/utils/frontend/frontend-utils.page.ts` at line 34:

```diff
async addToCart( itemName = '' ) {
  await this.page.waitForLoadState( 'domcontentloaded' );
  if ( itemName !== '' ) {
    await this.page
    .getByRole( 'button', {
-	name: `Add “${ itemName }” to your cart`,
+	name: `Add to cart: “${ itemName }”`,
      } )
      .click();
      } else {
	await this.page.click( 'text=Add to cart' );
      }
  // ...rest of the code
}
```

I also noticed that Woo Blocks `.env` file is missing, which causes some E2E test failures. You can recreate the `.env` file in `plugins/woocommerce-blocks/.env` (see the file in the Woo Blocks repo: https://github.com/woocommerce/woocommerce-blocks/blob/trunk/.env). Here is the content:

```
# WordPress container environment
WORDPRESS_DB_HOST=db
WORDPRESS_DB_NAME=wordpress
WORDPRESS_DB_USER=wordpress
WORDPRESS_DB_PASSWORD=wordpress

# WordPress CLI environment
WORDPRESS_PORT=8889
WORDPRESS_BASE_URL=http://localhost:8889
WORDPRESS_HOST=wordpress-www:80
WORDPRESS_TITLE=WooCommerce Core E2E Test Suite
WORDPRESS_LOGIN=admin
WORDPRESS_PASSWORD=password
WORDPRESS_EMAIL=admin@woocommercecoree2etestsuite.com
WP_VERSION=latest
WOO_VERSION=latest
```

Finally, I observed that detailed information on failing tests in CI is missing. In the Woo Blocks repository, we could view the total number of tests passed, failed, and skipped, along with specific errors for each failing test.

Commit:
Merge branch 'trunk' into dev/migrate-playwright-from-blocks

Commit:
Set up multiple test configurations

Commit:
Update failing test

Commit:
Update paths for artifacts

lanej0 commented:
Thanks for that review @tarhi-saad -- I've made the updates to the setup as described (and updated the one test). 

I'm not sure if somebody added it back, but [the `.env` file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/.env) is there now.

It looks like there's something going on with the artifacts upload for test results.  Probably another path that needs updating -- I'll get that resolved now.

Commit:
Update path to artifacts

Commit:
Path to results

tarhi-saad commented:
Thank you for the changes, @lanej0!

> It looks like there's something going on with the artifacts upload for test results. Probably another path that needs updating -- I'll get that resolved now.

The artifact is created in the Woo Blocks plugin, so maybe we write the whole path:

`path: plugins/woocommerce-blocks/tests/e2e/artifacts/${{ matrix.config.resultPath }}`

Two translation-related E2E tests are failing. When I ran them locally, I noticed that the Cart and Checkout pages weren't translated in the WordPress environment. However, this translation issue doesn't occur on a live site. I don't know what can cause the issue. I also made sure to update the translations in WP env. Can this be related to the setup?

gigitux committed:
remove unnecessary file

gigitux commented:
Thanks for working on this PR, @lanej0! 

> Two translation-related E2E tests are failing. When I ran them locally, I noticed that the Cart and Checkout pages weren't translated in the WordPress environment. However, this translation issue doesn't occur on a live site. I don't know what can cause the issue. I also made sure to update the translations in WP env. Can this be related to the setup?

I noticed that there were two files that contained the same test. The right file contains the suffix `.side_effects`. With https://github.com/woocommerce/woocommerce/pull/42804/commits/0ddc728e5d5d01a506c83924b4439ee57cfff284, I removed the other one.

gigitux commented:
@lanej0 I think that I understand what the issue is. Currently, in our E2E tests, we run some cli commands that use `npm`. For example, for the translations E2E tests, [we run](https://github.com/woocommerce/woocommerce/blob/0ddc728e5d5d01a506c83924b4439ee57cfff284/plugins/woocommerce-blocks/tests/e2e/tests/cart/cart-checkout-block-translations.shopper.block_theme.side_effects.spec.ts/#L23-L33):

```ts
	test.beforeAll( async () => {
		await cli(
			`npm run wp-env run tests-cli -- wp language core activate nl_NL`
		);
	} );

	test.afterAll( async () => {
		await cli(
			`npm run wp-env run tests-cli -- wp language core activate en_US`
		);
	} );
```

I guess that we should update them. What do you think?

tarhi-saad commented:
> I think that I understand what the issue is. Currently, in our E2E tests, we run some cli commands that use npm. For example, for the translations E2E tests

@gigitux, when I ran the tests in debug mode, I confirmed that the language was indeed switched, but the translation of our Blocks was missing.

I also opened the WP env site (http://localhost:8889), changed the language to `Netherlands`, and updated the translations manually, but the Cart and Checkout were still not translated.

tarhi-saad commented:
@lanej0, I'll skip the failing translation tests. We may have lost some translations after updating the domain (see discussion p1702648523959049-slack-C02UBB1EPEF). I'll create a ticket to investigate the translation issues!

I'll also test the artifact path that I suggested earlier to see if it works!

tarhi-saad committed:
Skip translation E2E tests

Skipping translation tests until we fix the missing translations issues after we changed the domain to "woocommerce"

tarhi-saad committed:
Update the artifacts path

tarhi-saad committed:
Fix lint errors

tarhi-saad reviewed (approved):
This looks good to me, @lanej0! I think it's ready to be merged!

lanej0 commented:
Thank you both for your work on this!
